### DISCUSSION TITLE
Channel Hopping with Crazyradio PA & Crazyflie

### DISCUSSION THREAD

**Discussion by MichaelZhou05** [bitcraze/discussions#816] [General]
Hi! last summer I was working on implementing a reinforcement learning based channel hopping algorithm on the Crazyflie 2.0. I was able to successfully implement the channel hopping algorithm by sending custom packets from cfclient to the crazyflie using the Crazyradio PA. Now as I am coming back to the project this summer, I read on the blog about the new Crazyradio 2.0 being developed with a new low latency channel hopping protocol. Does anyone from the Crazyflie team have any information on the new protocol or what improvements the new radio may provide?

**Comment by ataffanel** [comment]
Interesting project, I would be curious to know how well channel hopping was working with Crazyradio PA, did you modify the firmware or did you do everything over USB?

I am the main author of Crazyradio (pa and 2.0) firmware and communication code.

The main point of Crazyradio 2.0 is that it is based on the nRF52840, it is a Cortex-M4 with a lot of ram and flash and with a quite flexible radio hardware (similar to the one already present in the Crazyflie's nRF51). This means that it is much easier to program than the Crazyradio PA's 8051 and it allows much more control of the radio transmission scheduling. As of today though, the only published firmware for Crazyradio 2.0 is implementing a compatibility mode that makes it appears as a Crazyradio PA. The firmware project has another mode available that will allow the implementation of more protocols.

As for the channel hopping protocol, I currently mostly have a spec but no implementation yet. I have designed a protocol that looks very much like the BLE link layer but with a connection event period of 1ms (to achieve low latency as possible) and with fixed parameters (to simplify implementation). This protocol would be fully implemented in the Crazyradio 2.0 and so the lib would only see 2 data channel, up-link and down-link. This will make the lib much simpler but should also improve performance as we will not have so much traffic over USB anymore (no more data polling over usb for implementing the down-link).

So if you are working on the radio I suggest you get a Crazyradio 2.0. It has a debug port that is compatible with the Crazyflie so the Debug-adapter-kit allows to connect a SWD debug probe to it for debugging. It also has a safe bootloader so even without debugger it is much better to work with than Crazyradio PA.

If you want more info about our channel hopping plan, I could publish my draft specification for the protocol.

**Reply by MichaelZhou05** [reply]
Hi, thank you for your response!

I did not make any changes to the Crazyradio PA firmware, all was done through the Crazyflie python lib. All I did was send a custom packet from Crazyradio PA, then the slightly modified the firmware of the Crazyflie nRF receives and changes its channel. The Crazyradio PA would then switch to the new Channel, send an empty packet and wait for ack. This worked really well for me when testing the Channel Hopping algorithms on the ground. However, I do experience significant delay when trying to switch channels in flight. 

From my understanding of your blogs and response (sorry my limited understanding, still in highschool 😄 ) the main reason for the delays when switching Channels Mid flight (compared to when the Crazyflie is stationary) is due to the poor radio transmission scheduling of the Crazyradio PA's 8051 chip?

**Reply by ataffanel** [reply]
If I understand you well, you are sending a packet to the Crazyflie to tell it to change channel. This may be the cause of the problem since packet loss on one channel will make this channel change hard to achieve. My assumption is that the delay you observe are caused by some channel behaving very badly in flight and you get stuck on those.

A more common way to achieve channel hopping is with time: both side agree in advance to switch channel at regular interval. The hopping sequence is also known by both sides (it could be as simple as using channel+1 but it is often a pseudo-random sequence). The trick though is that requires both side to run exactly at the same speed in order to switch channel at the same time, and this is where firmware implementation and low level access to the radio helps: it allows us to synchronize hardware timers on both sides to achieve this synchronous channel hopping. I am simplifying a bit here, if you want (way more) details the [Bluetooth core specification](https://www.bluetooth.com/specifications/specs/core-specification-5-3/) is free to download, the BLE channel hopping scheme is described Vol6, PartB, 4.5.1.

If you want to continue working on the Python side, Crazyradio 2 is not going to help much, it currently does behave like a Crazyradio PA from the PC side. Though I would suggest investigating time-slot based channel hopping: if from USB you might have to go a bit slow, like one hop every 10-50ms. You can also be a bit creative by always trying to communicate with the current and the next channel, and have the Crazyflie switching channel automatically, this way you can detect when the channel hopping has happened.
